File: nodejs/src/token.ts
import type { Account, PublicClient, Transport, WalletClient } from "viem"
import type { base } from "viem/chains"
import { AFI_ADDRESS, ERC20_ABI } from "./constants.js"
import { ApprovalError, InsufficientBalanceError } from "./errors.js"
import { fetchTokenInfo } from "./multicall.js"
import { applyGasBuffer, feeFromReceipt } from "./swap.js"
import type { Address, PendingTx, TxReceipt, WaitForTxOptions } from "./types.js"

type BasePublicClient = PublicClient<Transport, typeof base>
type BaseWalletClient = WalletClient<Transport, typeof base, Account>

export async function getDecimals(token: Address, client: BasePublicClient): Promise<number> {
  return client.readContract({ address: token, abi: ERC20_ABI, functionName: "decimals" })
}

export async function getBalance(
  token: Address,
  owner: Address,
  client: BasePublicClient,
): Promise<bigint> {
  return client.readContract({
    address: token,
    abi: ERC20_ABI,
    functionName: "balanceOf",
    args: [owner],
  })
}

export async function getAllowance(
  token: Address,
  owner: Address,
  client: BasePublicClient,
): Promise<bigint> {
  return client.readContract({
    address: token,
    abi: ERC20_ABI,
    functionName: "allowance",
    args: [owner, AFI_ADDRESS],
  })
}

/** Reads `allowance(owner, spender)` against an arbitrary spender. */
export async function getAllowanceFor(
  token: Address,
  owner: Address,
  spender: Address,
  client: BasePublicClient,
): Promise<bigint> {
  return client.readContract({
    address: token,
    abi: ERC20_ABI,
    functionName: "allowance",
    args: [owner, spender],
  })
}

export async function assertSufficientBalance(
  token: Address,
  owner: Address,
  required: bigint,
  client: BasePublicClient,
): Promise<void> {
  const balance = await getBalance(token, owner, client)
  if (balance >= required) return
  // Enrich the error with symbol/decimals via multicall — one extra RPC on the failure path
  // gives a vastly better message ("Insufficient USDC: have 0.5, need 1") than raw addresses.
  const info = await fetchTokenInfo(token, client).catch(() => undefined)
  throw new InsufficientBalanceError(token, balance, required, owner, info?.symbol, info?.decimals)
}

export async function writeApprove(
  token: Address,
  spender: Address,
  amount: bigint,
  publicClient: BasePublicClient,
  walletClient: BaseWalletClient,
  account: Address,
  gasBufferPercent: number,
  nonce?: number,
): Promise<`0x${string}`> {
  const nonceParam = nonce !== undefined ? { nonce } : {}
  let gas: bigint
  try {
    gas = await publicClient.estimateContractGas({
      address: token,
      abi: ERC20_ABI,
      functionName: "approve",
      args: [spender, amount],
      account,
    })
  } catch {
    // Fall back to letting writeContract estimate again (some RPCs differ between estimate paths).
    return walletClient.writeContract({
      address: token,
      abi: ERC20_ABI,
      functionName: "approve",
      args: [spender, amount],
      ...nonceParam,
    })
  }
  return walletClient.writeContract({
    address: token,
    abi: ERC20_ABI,
    functionName: "approve",
    args: [spender, amount],
    gas: applyGasBuffer(gas, gasBufferPercent),
    ...nonceParam,
  })
}

/**
 * Ensures the AFI contract has exactly `amount` allowance for `token`.
 * Returns the approval tx hash, or null if existing allowance was already sufficient.
 */
export async function ensureExactApproval(
  token: Address,
  owner: Address,
  amount: bigint,
  publicClient: BasePublicClient,
  walletClient: BaseWalletClient,
  gasBufferPercent: number,
): Promise<string | null> {
  const current = await getAllowance(token, owner, publicClient)
  if (current >= amount) return null

  // Some tokens (USDT-style) reject non-zero → non-zero allowance changes. Reset first.
  // Preserve the reset error so it can be surfaced if the subsequent approve also fails.
  let resetErr: Error | undefined
  if (current > 0n) {
    try {
      const resetHash = await writeApprove(token, AFI_ADDRESS, 0n, publicClient, walletClient, owner, gasBufferPercent)
      await publicClient.waitForTransactionReceipt({ hash: resetHash })
    } catch (e) {
      resetErr = e as Error
    }
  }

  let hash: `0x${string}`
  try {
    hash = await writeApprove(token, AFI_ADDRESS, amount, publicClient, walletClient, owner, gasBufferPercent)
  } catch (e) {
    const msg = (e as Error).message
    throw new ApprovalError(resetErr ? `${msg} (allowance reset also failed: ${resetErr.message})` : msg)
  }

  await publicClient.waitForTransactionReceipt({ hash })

  // Verify on-chain before proceeding — protects against race conditions
  const confirmed = await getAllowance(token, owner, publicClient)
  if (confirmed < amount) {
    throw new ApprovalError("allowance not reflected on-chain after confirmation")
  }

  return hash
}

/**
 * Sends the approve tx and returns a PendingTx without waiting for confirmation.
 * Returns null if existing allowance is already sufficient.
 * The USDT-style reset pre-step still waits internally before sending the real approval.
 */
export async function submitApproval(
  token: Address,
  owner: Address,
  amount: bigint,
  publicClient: BasePublicClient,
  walletClient: BaseWalletClient,
  gasBufferPercent: number,
  nonce?: number,
): Promise<PendingTx | null> {
  const current = await getAllowance(token, owner, publicClient)
  if (current >= amount) return null

  let resetErr: Error | undefined
  let curNonce = nonce
  if (current > 0n) {
    try {
      const resetHash = await writeApprove(token, AFI_ADDRESS, 0n, publicClient, walletClient, owner, gasBufferPercent, curNonce)
      await publicClient.waitForTransactionReceipt({ hash: resetHash })
      if (curNonce !== undefined) curNonce += 1
    } catch (e) { resetErr = e as Error }
  }

  let hash: `0x${string}`
  try {
    hash = await writeApprove(token, AFI_ADDRESS, amount, publicClient, walletClient, owner, gasBufferPercent, curNonce)
  } catch (e) {
    const msg = (e as Error).message
    throw new ApprovalError(resetErr ? `${msg} (allowance reset also failed: ${resetErr.message})` : msg)
  }

  return {
    txHash: hash,
    wait: async (opts?: WaitForTxOptions): Promise<TxReceipt> => {
      const receipt = await publicClient.waitForTransactionReceipt({
        hash,
        confirmations: opts?.confirmations,
        timeout: opts?.timeoutMs,
      })
      const confirmed = await getAllowance(token, owner, publicClient)
      if (confirmed < amount) throw new ApprovalError("allowance not reflected on-chain after confirmation")
      const fee = feeFromReceipt(receipt.gasUsed, (receipt as { effectiveGasPrice?: bigint }).effectiveGasPrice)
      return { blockNumber: receipt.blockNumber, gasUsed: receipt.gasUsed, ...fee }
    },
  }
}

File: go/token.go
package afi

import (
	"context"
	"fmt"
	"math/big"
	"strings"

	"github.com/ethereum/go-ethereum"
	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/ethclient"
)

func callERC20(ctx context.Context, client *ethclient.Client, erc20ABI abi.ABI, token common.Address, method string, args ...any) ([]any, error) {
	input, err := erc20ABI.Pack(method, args...)
	if err != nil {
		return nil, fmt.Errorf("pack %s: %w", method, err)
	}

	result, err := client.CallContract(ctx, ethereum.CallMsg{To: &token, Data: input}, nil)
	if err != nil {
		return nil, fmt.Errorf("call %s: %w", method, err)
	}

	return erc20ABI.Unpack(method, result)
}

func getDecimals(ctx context.Context, client *ethclient.Client, erc20ABI abi.ABI, token common.Address) (uint8, error) {
	out, err := callERC20(ctx, client, erc20ABI, token, "decimals")
	if err != nil {
		return 0, err
	}
	return out[0].(uint8), nil
}

func getBalance(ctx context.Context, client *ethclient.Client, erc20ABI abi.ABI, token, owner common.Address) (*big.Int, error) {
	out, err := callERC20(ctx, client, erc20ABI, token, "balanceOf", owner)
	if err != nil {
		return nil, err
	}
	return out[0].(*big.Int), nil
}

func getAllowance(ctx context.Context, client *ethclient.Client, erc20ABI abi.ABI, token, owner common.Address) (*big.Int, error) {
	return getAllowanceFor(ctx, client, erc20ABI, token, owner, AfiAddress)
}

// getAllowanceFor reads ERC20(token).allowance(owner, spender). Generic version
// of getAllowance — required by workflow prechecks that target the NMR contract
// rather than the AFI router.
func getAllowanceFor(ctx context.Context, client *ethclient.Client, erc20ABI abi.ABI, token, owner, spender common.Address) (*big.Int, error) {
	out, err := callERC20(ctx, client, erc20ABI, token, "allowance", owner, spender)
	if err != nil {
		return nil, err
	}
	return out[0].(*big.Int), nil
}

func assertSufficientBalance(ctx context.Context, client *ethclient.Client, erc20ABI, mcABI abi.ABI, token, owner common.Address, required *big.Int) error {
	balance, err := getBalance(ctx, client, erc20ABI, token, owner)
	if err != nil {
		return err
	}
	if balance.Cmp(required) >= 0 {
		return nil
	}
	// Enrich the error with symbol/decimals via multicall — one extra RPC on the failure path
	// gives a vastly better message than raw addresses.
	symbol, decimals := "", uint8(0)
	if info, err := fetchTokenInfo(ctx, client, erc20ABI, mcABI, token, common.Address{}, nil); err == nil {
		symbol = info.Symbol
		decimals = info.Decimals
	}
	return ErrInsufficientBalanceDetailed(token.Hex(), owner.Hex(), symbol, decimals, balance, required)
}

// submitRevoke sends approve(AFI, 0) for `token`. Returns nil when allowance is already zero.
func submitRevoke(
	ctx context.Context,
	client *ethclient.Client,
	erc20ABI abi.ABI,
	c *Client,
	token, owner common.Address,
) (*PendingTx, error) {
	current, err := getAllowance(ctx, client, erc20ABI, token, owner)
	if err != nil {
		return nil, err
	}
	if current.Sign() == 0 {
		return nil, nil
	}

	hash, err := sendApproveTx(ctx, c, erc20ABI, token, big.NewInt(0))
	if err != nil {
		return nil, ErrApproval(err.Error())
	}

	pending := &PendingTx{
		TxHash: hash,
		waitFn: func(ctx context.Context, opts WaitForTxOptions) (*TxReceipt, error) {
			receipt, err := c.waitReceiptWithOpts(ctx, hash, opts)
			if err != nil {
				return nil, err
			}
			confirmed, err := getAllowance(ctx, client, erc20ABI, token, owner)
			if err != nil {
				return nil, err
			}
			if confirmed.Sign() != 0 {
				return nil, ErrApproval("allowance not zeroed after confirmation")
			}
			return receiptToTxReceipt(receipt), nil
		},
	}
	return pending, nil
}

// fetchTokenInfo runs symbol+name+decimals (+balance+allowance when owner != zero) in a single multicall.
// `cache` is consulted for metadata first and populated on misses.
func fetchTokenInfo(
	ctx context.Context,
	client *ethclient.Client,
	erc20ABI, mcABI abi.ABI,
	token, owner common.Address,
	cache *metadataCache,
) (*TokenInfo, error) {
	infos, err := fetchTokenInfoBatch(ctx, client, erc20ABI, mcABI, []common.Address{token}, owner, cache)
	if err != nil {
		return nil, err
	}
	return infos[0], nil
}

// fetchTokenInfoBatch is the N-token form of fetchTokenInfo — one multicall round-trip.
// Tokens whose metadata is already in `cache` only fetch balance/allowance (or nothing, when no owner).
func fetchTokenInfoBatch(
	ctx context.Context,
	client *ethclient.Client,
	erc20ABI, mcABI abi.ABI,
	tokens []common.Address,
	owner common.Address,
	cache *metadataCache,
) ([]*TokenInfo, error) {
	if len(tokens) == 0 {
		return []*TokenInfo{}, nil
	}
	withOwner := owner != (common.Address{})

	pack := func(method string, args ...any) []byte {
		data, err := erc20ABI.Pack(method, args...)
		if err != nil {
			panic(fmt.Errorf("pack %s: %w", method, err))
		}
		return data
	}

	type plan struct {
		token  common.Address
		cached *TokenMetadata
		offset int
		count  int
	}

	plans := make([]plan, 0, len(tokens))
	calls := make([]Multicall3Call, 0, len(tokens)*5)
	for _, t := range tokens {
		var cached *TokenMetadata
		if m, ok := cache.get(t); ok {
			cached = &m
		}
		p := plan{token: t, cached: cached, offset: len(calls)}
		if cached == nil {
			calls = append(calls,
				Multicall3Call{Target: t, AllowFailure: true, CallData: pack("symbol")},
				Multicall3Call{Target: t, AllowFailure: true, CallData: pack("name")},
				Multicall3Call{Target: t, AllowFailure: false, CallData: pack("decimals")},
			)
			p.count += 3
		}
		if withOwner {
			calls = append(calls,
				Multicall3Call{Target: t, AllowFailure: false, CallData: pack("balanceOf", owner)},
				Multicall3Call{Target: t, AllowFailure: false, CallData: pack("allowance", owner, AfiAddress)},
			)
			p.count += 2
		}
		plans = append(plans, p)
	}

	results, err := aggregate3(ctx, client, mcABI, calls)
	if err != nil {
		return nil, err
	}

	out := make([]*TokenInfo, len(tokens))
	for i, p := range plans {
		info := &TokenInfo{Address: p.token, Owner: owner}
		cursor := p.offset

		var meta TokenMetadata
		if p.cached != nil {
			meta = *p.cached
		} else {
			if results[cursor].Success {
				if dec, err := erc20ABI.Unpack("symbol", results[cursor].ReturnData); err == nil && len(dec) > 0 {
					if s, ok := dec[0].(string); ok {
						meta.Symbol = s
					}
				}
			}
			cursor++
			if results[cursor].Success {
				if dec, err := erc20ABI.Unpack("name", results[cursor].ReturnData); err == nil && len(dec) > 0 {
					if s, ok := dec[0].(string); ok {
						meta.Name = s
					}
				}
			}
			cursor++
			dec, err := erc20ABI.Unpack("decimals", results[cursor].ReturnData)
			if err != nil {
				return nil, fmt.Errorf("unpack decimals for %s: %w", p.token.Hex(), err)
			}
			meta.Decimals = dec[0].(uint8)
			cursor++
			cache.set(p.token, meta)
		}

		info.Symbol = meta.Symbol
		info.Name = meta.Name
		info.Decimals = meta.Decimals

		if withOwner {
			bal, err := erc20ABI.Unpack("balanceOf", results[cursor].ReturnData)
			if err != nil {
				return nil, fmt.Errorf("unpack balanceOf for %s: %w", p.token.Hex(), err)
			}
			info.Balance = bal[0].(*big.Int)
			cursor++
			al, err := erc20ABI.Unpack("allowance", results[cursor].ReturnData)
			if err != nil {
				return nil, fmt.Errorf("unpack allowance for %s: %w", p.token.Hex(), err)
			}
			info.Allowance = al[0].(*big.Int)
		}

		out[i] = info
	}

	return out, nil
}

// ensureExactApproval approves exactly `amount` for the AFI contract.
// Returns the approval tx hash, or empty string if allowance was already sufficient.
func ensureExactApproval(
	ctx context.Context,
	client *ethclient.Client,
	erc20ABI abi.ABI,
	c *Client,
	token, owner common.Address,
	amount *big.Int,
) (string, error) {
	current, err := getAllowance(ctx, client, erc20ABI, token, owner)
	if err != nil {
		return "", err
	}

	if current.Cmp(amount) >= 0 {
		return "", nil
	}

	// Some tokens (USDT-style) reject non-zero → non-zero allowance changes.
	// Reset to 0 first if needed. We preserve the reset error so it can be surfaced
	// when the subsequent approve also fails — otherwise the dev sees a confusing
	// "allowance not reflected" without knowing the reset was the real culprit.
	var resetErr error
	if current.Sign() > 0 {
		if _, err := sendApprove(ctx, c, erc20ABI, token, big.NewInt(0)); err != nil {
			resetErr = err
		}
	}

	hash, err := sendApprove(ctx, c, erc20ABI, token, amount)
	if err != nil {
		if resetErr != nil {
			return "", ErrApproval(fmt.Sprintf("%s (allowance reset also failed: %s)", err.Error(), resetErr))
		}
		return "", ErrApproval(err.Error())
	}

	// Verify allowance on-chain before proceeding
	confirmed, err := getAllowance(ctx, client, erc20ABI, token, owner)
	if err != nil {
		return "", err
	}
	if confirmed.Cmp(amount) < 0 {
		return "", ErrApproval("allowance not reflected on-chain after confirmation")
	}

	return hash, nil
}

func sendApproveTx(ctx context.Context, c *Client, erc20ABI abi.ABI, token common.Address, amount *big.Int) (string, error) {
	return sendApproveTxWithNonce(ctx, c, erc20ABI, token, amount, nil)
}

func sendApproveTxWithNonce(ctx context.Context, c *Client, erc20ABI abi.ABI, token common.Address, amount *big.Int, nonce *uint64) (string, error) {
	input, err := erc20ABI.Pack("approve", AfiAddress, amount)
	if err != nil {
		return "", err
	}
	return c.sendTx(ctx, &token, input, c.gasBufferPct, nonce)
}

func sendApprove(ctx context.Context, c *Client, erc20ABI abi.ABI, token common.Address, amount *big.Int) (string, error) {
	hash, err := sendApproveTx(ctx, c, erc20ABI, token, amount)
	if err != nil {
		return "", err
	}
	_, err = c.waitReceipt(ctx, hash)
	if err != nil {
		return "", err
	}
	return strings.ToLower(hash), nil
}

func submitApprove(
	ctx context.Context,
	client *ethclient.Client,
	erc20ABI abi.ABI,
	c *Client,
	token, owner common.Address,
	amount *big.Int,
) (*PendingTx, error) {
	current, err := getAllowance(ctx, client, erc20ABI, token, owner)
	if err != nil {
		return nil, err
	}
	if current.Cmp(amount) >= 0 {
		return nil, nil
	}

	var resetErr error
	if current.Sign() > 0 {
		if _, err := sendApprove(ctx, c, erc20ABI, token, big.NewInt(0)); err != nil {
			resetErr = err
		}
	}

	hash, err := sendApproveTx(ctx, c, erc20ABI, token, amount)
	if err != nil {
		if resetErr != nil {
			return nil, ErrApproval(fmt.Sprintf("%s (allowance reset also failed: %s)", err.Error(), resetErr))
		}
		return nil, ErrApproval(err.Error())
	}

	amountCopy := new(big.Int).Set(amount)
	pending := &PendingTx{
		TxHash: hash,
		waitFn: func(ctx context.Context, opts WaitForTxOptions) (*TxReceipt, error) {
			receipt, err := c.waitReceiptWithOpts(ctx, hash, opts)
			if err != nil {
				return nil, err
			}
			confirmed, err := getAllowance(ctx, client, erc20ABI, token, owner)
			if err != nil {
				return nil, err
			}
			if confirmed.Cmp(amountCopy) < 0 {
				return nil, ErrApproval("allowance not reflected on-chain after confirmation")
			}
			return receiptToTxReceipt(receipt), nil
		},
	}
	return pending, nil
}
